# Исследование надежности заемщиков

Заказчик — кредитный отдел банка. Нам нужно разобраться: влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты нашего исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Описание данных**
- children — количество детей в семье;
- days_employed — общий трудовой стаж в днях;
- dob_years — возраст клиента в годах;
- education — уровень образования клиента;
- education_id — идентификатор уровня образования;
- family_status — семейное положение;
- family_status_id — идентификатор семейного положения;
- gender — пол клиента;
- income_type — тип занятости;
- debt — имел ли задолженность по возврату кредитов;
- total_income — ежемесячный доход;
- purpose — цель получения кредита.

## Изучение общей информации о данных

Импортируем библиотеку pandas. Прочтем данные из файла и сохраним в переменную data. 

In [1]:
import pandas as pd
# чтение файла и сохранение в переменную data
data = pd.read_csv('data.csv')

Выведем первые 20 строчек датафрейма data на экран.

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Выведем основную информацию о датафрейме с помощью метода info().

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод**

В исходных данных 12 колонок и 21 525 строк с информацией о заемщиках. Из них 2 столбца типа float64, 5 столбцов типа int64 и 5 столбцов типа object. В столбцах days_employed и total_income есть пропуски.

## Предобработка данных

### Обработка аномальных значений

В столбце days_employed встречается отрицательное количество дней трудового стажа. Вероятно такая аномалия возникла по какой-то ошибке. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода abs().

In [4]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа days_employed в днях.

In [5]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получились аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится нам для дальнейшего исследования.

Посмотрим уникальные значения столбца children.

In [6]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце children есть два аномальных значения. Удалим строки, в которых встречаются эти аномальные значения из датафрейма data.

In [7]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выведем уникальные значения столбца children, чтобы убедиться, что аномалии удалены.

In [8]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца.

In [9]:
data.isna().sum()

children               0
days_employed       2162
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2162
purpose                0
dtype: int64

В столбцах days_employed и total_income есть пропуски. Столбец total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type.

In [10]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type.

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Проверим, что все пропуски заполнены. 

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце total_income на целочисленный с помощью метода astype().

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

В столбце education есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [14]:
data['education'] = data['education'].str.lower()

Проверим датафрейм data на наличие явных дубликатов.

In [15]:
data.duplicated().sum()

71

Удалим эти дубликаты.

In [16]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов доходов кредитополучателей, указанных ниже, создадим в датафрейме data столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass
    
# добавим столбец total_income_category
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран перечень уникальных целей взятия кредита из столбца purpose.

In [18]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [19]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'
    
# добавим столбец purpose_category
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных

Посмотрим: есть ли зависимость между количеством детей и возвратом кредита в срок.

In [20]:
children_debt = data.pivot_table(index='children', values='debt', aggfunc='mean').reset_index()
children_debt.columns = ['Количество детей', 'Процент должников']
children_debt['Процент должников'] = round(children_debt['Процент должников'] * 100, 2)
children_debt

Количество детей  Процент должников
0                 0               7.54
1                 1               9.23
2                 2               9.45
3                 3               8.18
4                 4               9.76
5                 5               0.00

**Вывод**

С учетом этой таблицы, можно сделалть вывод, что если у людей 5 и более детей, то процент должников равен 0. Но если брать в общей массе людей по городу (региону, стране проживания), то очень маленький процент людей имеют 5 и более детей. Поэтому лучше рассматривать людей, у котороых от 1 до 4 детей включительно, и у кого нет детей. А согласно таблице выше количество детей не влияет на возврат кредита в срок. Влияет в основном наличие детей у людей. Процент должников без детей составляет 7,54. В тоже время процент должников, имеющих одного, двух или четырех детей, - соответсвенно 9,23, 9,45 и 9,76. Немного ниже процент должников, у кого трое детей (8,18).

Посмотрим: есть ли зависимость между семейным положением и возвратом кредита в срок.

In [21]:
family_status_debt = data.pivot_table(index='family_status', columns='debt', values='purpose', aggfunc='count').reset_index()
family_status_debt['ratio'] = round(family_status_debt[1] / (family_status_debt[0] + family_status_debt[1]) * 100, 2)
family_status_debt.columns =['Семейное положение', 'Не имеют долга', 'Имеют долг', 'Доля должников']
family_status_debt

Семейное положение  Не имеют долга  Имеют долг  Доля должников
0  Не женат / не замужем            2523         273            9.76
1              в разводе            1105          84            7.06
2         вдовец / вдова             888          63            6.62
3       гражданский брак            3749         385            9.31
4        женат / замужем           11334         927            7.56

**Вывод**

Доля должников, которые не состоят в семейных отношениях либу живут в гражданском браке, составляет от 9.3% до 9.8%. В то время как доля должников, состоящих в семейных отношениях либо разведены, - от 6.6% до 7.6%. Следовательно, семейное положение влияет на возврат кредита в срок.

Чтобы понять: влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок, сделаем таблицу. В таблице укажем семейное положение и наличие детей клиента, а также долю должников по каждой категории.

In [22]:
#создадим столбец о наличии детей, где 0 - нет детей, 1 - есть дети
def categorize_children(children):
    try:
        if children == 0:
            return 0
        if 1 <= children <= 5:
            return 1
    except:
        pass
    
# добавим столбец children_category
data['children_category'] = data['children'].apply(categorize_children)

In [23]:
#создадим сводную таблицу, где покажем семейное положение, наличие детей, долю должников в этих категориях
family_status_children_debt = data.pivot_table(
    index=['family_status', 'children_category'], columns='debt', values='purpose', aggfunc='count').reset_index()
family_status_children_debt['ratio'] = round(
    family_status_children_debt[1] / (family_status_children_debt[0] + family_status_children_debt[1]) * 100, 2)
family_status_children_debt.columns =['Семейное положение', 'Наличие детей', 'Не имеют долга', 'Имеют долг', 'Доля должников']
family_status_children_debt

Семейное положение  Наличие детей  Не имеют долга  Имеют долг  \
0  Не женат / не замужем              0            2052         210   
1  Не женат / не замужем              1             471          63   
2              в разводе              0             729          55   
3              в разводе              1             376          29   
4         вдовец / вдова              0             794          53   
5         вдовец / вдова              1              94          10   
6       гражданский брак              0            2501         229   
7       гражданский брак              1            1248         156   
8        женат / замужем              0            6952         516   
9        женат / замужем              1            4382         411   

   Доля должников  
0            9.28  
1           11.80  
2            7.02  
3            7.16  
4            6.26  
5            9.62  
6            8.39  
7           11.11  
8            6.91  
9            8.58

**Вывод**

Как было выше сказано само количество детей не влияет на факт погашения кредита, а вот налие детей влияет. Среди не женатых / не замужем людей без детей доля должников составляет 9.3% в то время, как доля должников не женатых/не замужем людей с детьми - 11.7%. Похожая ситуация наблюдается и у людей, живущих гражданским браком: доля должников без детей - 8.3%, доля должников с детьми - 11,2%. Среди людей, зарегистрировавших свой брак, доля должников без детей составляет 6.9%, с детьми - 8.6%. Среди овдовевших людей без детей доля должников - 6.2%, а с детьми - 9.6%. В категории разведенных людей доля должников без детей практически не отличается от доли должников с детьми (7.1% и 7.2% соответсвенно).

Посмотрим: есть ли зависимость между уровнем дохода и возвратом кредита в срок.

In [24]:
total_income_debt = data.pivot_table(
    index='total_income_category', columns='debt', values='purpose', aggfunc='count').reset_index()
total_income_debt['ratio'] = round(total_income_debt[1] / (total_income_debt[0] + total_income_debt[1]) * 100, 2)
total_income_debt.columns = ['Категория ежемесячного дохода', 'Не имеют долга', 'Имеют долг', 'Доля должников']
total_income_debt

Категория ежемесячного дохода  Не имеют долга  Имеют долг  Доля должников
0                             A              23           2            8.00
1                             B            4660         354            7.06
2                             C           14568        1353            8.50
3                             D             328          21            6.02
4                             E              20           2            9.09

**Вывод**

Если сравнить самые многочисленные категории, то большая часть должников будет с доходом от 50 000 до 200 000 (8.5%), а меньше должников будет в категории с доходом от 200 000 до 1 000 000 (7.1%). То есть, на 1000 человек с доходом от 50 000 до 200 000 будет 93 должника, а на 1000 человек с доходом от 200 000 до 1 000 000 - 76 должников.

Посмотрим: как разные цели кредита влияют на его возврат в срок.

In [25]:
purpose_debt = data.pivot_table(index='purpose_category', columns='debt', values='purpose', aggfunc='count').reset_index()
purpose_debt['ratio'] = round(purpose_debt[1] / (purpose_debt[0] + purpose_debt[1]) * 100, 2)
purpose_debt.columns = ['Цель кредита', 'Не имеют долг', 'Имеют долг', 'Доля должников']
purpose_debt

Цель кредита  Не имеют долг  Имеют долг  Доля должников
0    операции с автомобилем           3879         400            9.35
1  операции с недвижимостью           9971         780            7.26
2     получение образования           3619         369            9.25
3        проведение свадьбы           2130         183            7.91

**Вывод** 

В основном задолжности по кредитам возникают, когда берут кредиты на операции с автомобилем или на образование. Доля должников по этим кредитам составляет 9.35% и 9.25% соответсвенно. Задолжность по кредитам на проведение свадьбы всего 7.9%. Меньше всего составляет задолжность по операциям с недвижимостью - всего 7.3%.

## Итоговый вывод

В результате исследования было выявлено, что само количество детей не влияет на факт погашения кредита, а вот налие детей влияет. Среди не женатых / не замужем людей без детей доля должников составляет 9.3% в то время, как доля должников не женатых/не замужем людей с детьми - 11.7%. Похожая ситуация наблюдается и у людей, живущих гражданским браком: доля должников без детей - 8.3%, доля должников с детьми - 11,2%. Среди людей, зарегистрировавших свой брак, доля должников без детей составляет 6.9%, с детьми - 8.6%. Среди овдовевших людей без детей доля должников - 6.2%, а с детьми - 9.6%. В категории разведенных людей доля должников без детей практически не отличается от доли должников с детьми (7.1% и 7.2% соответсвенно). Также основная доля должников по кредитам приходится на людей с доходом до 30 000 и людей с доходом от 50 000 до 200 000 (9.1 % и 8.5% соответсвенно). Меньше всего должников с ежемесячным доходом от 30 000 до 50 000 (всего 6%). Доля должников с доходом от 200 000 и выше составляет 7.1-8%. Если сравнить самые многочисленные категории по ежемесячным доходам, то большая часть должников будет с доходом от 50 000 до 200 000 (8.5%), а меньше должников будет в категории с доходом от 200 000 до 1 000 000 (7.1%). То есть, на 1000 человек с доходом от 50 000 до 200 000 будет 93 должника, а на 1000 человек с доходом от 200 000 до 1 000 000 - 76 должников. А еще задолжности по кредитам возникают, когда берут кредиты на операции с автомобилем или на образование. Доля должников по этим кредитам составляет 9.35% и 9.25% соответсвенно. Задолжность по кредитам на проведение свадьбы всего 7.9%. Меньше всего составляет задолжность по операциям с недвижимостью - всего 7.3%.